solidity/src/Polymer7683.sol

```diff
@@ -0,0 +1,208 @@
+    function handleRefundWithProof(
+        bytes32 orderId,
+        bytes calldata eventProof
+    ) external {
+        // 1. Use orderId for replay protection
+        if (processedEvents[orderId]) revert EventAlreadyProcessed();
+
+        // 2. Validate order ID is in the refunded set
+        bytes32[] memory eventOrderIds = _validateRefundProof(eventProof);
+        bool found = false;
+        for (uint256 i = 0; i < eventOrderIds.length; i++) {
+            if (eventOrderIds[i] == orderId) {
+                found = true;
+                break;
+            }
+        }
+        if (!found) revert InvalidEventData();
+
+        // 8. Process refund for the order
+        processedEvents[orderId] = true;
+        _handleRefundOrder(orderId);
+    }
+
+    // ============ Internal Functions ============
+
+    /**
+     * @notice Dispatches a settlement instruction by emitting a Filled event that will be proven on the origin chain
+     * @param _originDomain The domain to which the settlement message is sent
+     * @param _orderIds The IDs of the orders to settle
+     * @param _ordersFillerData The filler data for the orders
+     */
+    function _dispatchSettle(
+        uint32 _originDomain,
+        bytes32[] memory _orderIds,
+        bytes[] memory _ordersFillerData
+    ) internal override {
+        for (uint256 i = 0; i < _orderIds.length; i++) {
+            FilledOrder memory order = filledOrders[_orderIds[i]];
+            emit Filled(_orderIds[i], order.originData, order.fillerData);
+        }
+    }
+
+    /**
+     * @notice Dispatches a refund instruction by emitting a Refund event that will be proven on the origin chain
+     * @param _originDomain The domain to which the refund message is sent
+     * @param _orderIds The IDs of the orders to refund
+     */
+    function _dispatchRefund(
+        uint32 _originDomain,
+        bytes32[] memory _orderIds
+    ) internal override {
+        emit Refund(_orderIds);
+    }
+
+    /**
+     * @notice Retrieves the local domain identifier
+     * @return The local domain ID (chain ID)
+     */
+    function _localDomain() internal view override returns (uint32) {
+        return uint32(localChainId);
+    }
+
+    function _validateSettlementProof(
+        bytes calldata eventProof
+    ) private view returns (
+        bytes32 eventOrderId,
+        bytes memory originData,
+        bytes memory fillerData
+    ) {
+        (
+            uint32 provenChainId,
+            address actualEmitter,
+            ,  // topics
+            bytes memory data
+        ) = prover.validateEvent(eventProof);
+
+        // Decode settlement-specific data
+        (eventOrderId, originData, fillerData) = abi.decode(data, (bytes32, bytes, bytes));
+
+        // Validate chain ID with origin data
+        OrderData memory orderData = OrderEncoder.decode(originData);
+        if (provenChainId != orderData.destinationDomain) {
+            revert InvalidChainId();
+        }
```

Comment:
good catch!

Reply:
Thanks to this catch we realized there were a potential vulnerability in the our base contacts and we addressed it with this changes https://github.com/BootNodeDev/intents-framework/pull/112
Please notice the changes on both `_handleSettleOrder` and  `_handleRefundOrder` which now receceive the msg origin and msg sender of the settle and refund, which in you case would be the `provenChainId` and `actualEmitter` respectively. 
If I'm not wrong the fix should help simplify this and `_validateRefundProof`

Reply:
yes, i've rebased this PR on #112 and refactored the code to simplify.